test/qmedia.cpp

```diff
@@ -1,7 +1,268 @@
+        collector->qualityProfile = qualityProfile;
+        return 0;
+    }
+
+    int update(const std::string& /* sourceId */,
+               const std::string& /* label */,
+               const std::string& /* qualityProfile */) override
+    {
+        // Always error on update to force a prepare call
+        return 1;
+    }
+
+    int subscribedObject(quicr::bytes&& data, std::uint32_t groupId, std::uint16_t objectId) override
+    {
+        collector->add_object({ groupId, objectId, data });
+        return 0;
+    }
+
+private:
+    std::shared_ptr<SubscriptionCollector> collector;
+};
+
+class QSubscriberTestDelegate : public qmedia::QSubscriberDelegate
 {
-    REQUIRE(1 == 1);
+public:
+    QSubscriberTestDelegate(std::shared_ptr<SubscriptionCollector> collector_in) : collector(std::move(collector_in)) {}
+
+    virtual ~QSubscriberTestDelegate() = default;
+
+    std::shared_ptr<qmedia::QSubscriptionDelegate> allocateSubByNamespace(const quicr::Namespace& /* quicrNamespace */,
+                                                                          const std::string& /* qualityProfile */)
+    {
+        return std::make_shared<QSubscriptionTestDelegate>(collector);
+    }
+
+    int removeSubByNamespace(const quicr::Namespace& /* quicrNamespace */) { return 0; }
+
+private:
+    std::shared_ptr<SubscriptionCollector> collector;
+};
+
+class QPublicationTestDelegate : public qmedia::QPublicationDelegate
+{
+public:
+    virtual ~QPublicationTestDelegate() = default;
+
+    int prepare(const std::string& /* sourceId */, const std::string& /* qualityProfile */, bool& /* reliable */)
+    {
+        return 0;
+    }
+
+    int update(const std::string& /* sourceId */, const std::string& /* qualityProfile */) { return 0; }
+
+    void publish(bool /* pubFlag */) {}
+};
+
+class QPublisherTestDelegate : public qmedia::QPublisherDelegate
+{
+public:
+    virtual ~QPublisherTestDelegate() = default;
+
+    std::shared_ptr<qmedia::QPublicationDelegate> allocatePubByNamespace(const quicr::Namespace& /* quicrNamespace */,
+                                                                         const std::string& /* sourceID */,
+                                                                         const std::string& /* qualityProfile */)
+    {
+        return std::make_shared<QPublicationTestDelegate>();
+    }
+
+    int removePubByNamespace(const quicr::Namespace& /* quicrNamespace */) { return 0; }
+};
+
+static qmedia::QController make_controller(std::shared_ptr<SubscriptionCollector> collector)
+{
+    const auto sub = std::make_shared<QSubscriberTestDelegate>(std::move(collector));
+    const auto pub = std::make_shared<QPublisherTestDelegate>();
+    const auto logger = std::make_shared<cantina::Logger>("QTest", "QTEST");
+    return {sub, pub, logger};
+}
+
+static qmedia::manifest::MediaStream make_media_stream(uint32_t endpoint_id)
+{
+    const auto source_base = "source ";
+    const auto label_base = "Participant ";
```

Comment:
Do you mean these to be `const char*`? Do you want std::string instead?
```suggestion
    const std::string source_base = "source ";
    const std::string label_base = "Participant ";
```
Or, in defiance of [cppcoreguidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es11-use-auto-to-avoid-redundant-repetition-of-type-names) and [clang-tidy](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-auto.html):
```suggestion
    const auto source_base = std::string("source ");
    const auto label_base = std::string("Participant ");
```
```

Reply:
Actually this is a handy use for [`operator""s`](https://en.cppreference.com/w/cpp/string/basic_string/operator%22%22s)

Reply:
AAA failed :)  Seriously, I think it's generally better to be explicit. I think `auto` is fine for range-based for loops, creating iterators, defining a lambda, etc. But I just don't agree with use where it's easy to be explicit.